pkg/pdp/scheduler/handler.go

```diff
@@ -0,0 +1,265 @@
+package scheduler
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	"gorm.io/gorm"
+
+	"github.com/storacha/storage/pkg/database"
+	"github.com/storacha/storage/pkg/pdp/service/models"
+)
+
+// taskTypeHandler ties a task implementation with engine-specific metadata.
+type taskTypeHandler struct {
+	TaskInterface
+	TaskTypeDetails TaskTypeDetails
+	TaskEngine      *TaskEngine
+	doneMu          sync.Mutex
+}
+
+// AddTask is the implementation passed to each task's Adder.
+// It creates a new task record in the database.
+func (h *taskTypeHandler) AddTask(extra func(TaskID, *gorm.DB) (bool, error)) {
+	var tID TaskID
+	retryWait := 100 * time.Millisecond
+
+retryAddTask:
+	err := h.TaskEngine.db.WithContext(h.TaskEngine.ctx).Transaction(func(tx *gorm.DB) error {
+		// Create a new Task record.
+		task := models.Task{
+			PostedTime: time.Now(),
+			AddedBy:    h.TaskEngine.sessionID,
+			Name:       h.TaskTypeDetails.Name,
+		}
+
+		// Insert the task and let GORM fill in the auto-generated ID.
+		if err := tx.Create(&task).Error; err != nil {
+			return fmt.Errorf("could not insert task: %w", err)
+		}
+
+		// Set the task ID from the newly inserted record.
+		tID = TaskID(task.ID)
+
+		// Call the extra callback to update additional info in the same transaction.
+		shouldCommit, err := extra(tID, tx)
+		if err != nil {
+			return err
+		}
+
+		if shouldCommit {
+			return nil
+		}
+
+		return ErrDoNotCommit
```

Comment:
Returning an error that is ignored below made this difficult for me to understand. Was there some functionality that was removed that would have been done when `shouldCommit` is true? Perhaps this needs a comment explaining?

Reply:
Yeah, this is _really_ confusing, I'll add a comment here. To explain:
- `AddTask` attempts to add a task to the database.
- `AddTask` invokes a tasks `Adder` function (`extra` in this method). The `Adder` function, implemented by each tasks, is used to determine if said task should run.
- If the task needs to run, its `Adder` returns `true`, otherwise `false`.
- In the event the `Adder` returns `false` we don't want to run the task, so we return an error here to abort the database transactions which creates the task in the db. 

I'll add a comment to clarify this.